gateway/src/main/kotlin/handler/HandshakeHandler.kt

```diff
@@ -19,31 +19,37 @@ internal class HandshakeHandler(
 
     lateinit var configuration: GatewayConfiguration
 
-    private val session: AtomicRef<String?> = atomic(null)
+    // see https://discord.com/developers/docs/topics/gateway#resuming
+    private class ResumeContext(val sessionId: String, val resumeUrl: Url)
 
-    private val identify
-        get() = configuration.identify
+    private val resumeContext = atomic<ResumeContext?>(initial = null)
+    val gatewayUrl get() = resumeContext.value?.resumeUrl ?: initialUrl
 
-    private val resume
-        get() = Resume(configuration.token, session.value!!, sequence.value ?: 0)
-
-    private val sessionStart get() = session.value == null
-
-    override fun start() {
-        on<Ready> { event ->
-            session.update { event.data.sessionId }
+    private val resumeOrIdentify
+        get() = when (val sessionId = resumeContext.value?.sessionId) {
+            null -> configuration.identify
+            else -> Resume(configuration.token, sessionId, sequence.value ?: 0)
         }
 
+    override fun start() {
         on<Hello> {
-            reconnectRetry.reset() //connected and read without problems, resetting retry counter
+            reconnectRetry.reset() // connected and read without problems, resetting retry counter
             identifyRateLimiter.consume {
-                if (sessionStart) send(identify)
-                else send(resume)
+                send(resumeOrIdentify)
             }
         }
 
+        on<Ready> { event ->
+            // keep custom query params
+            val resumeUrl = URLBuilder(event.data.resumeGatewayUrl)
+                .apply { parameters.appendMissing(initialUrl.parameters) }
+                .build()
+
+            resumeContext.value = ResumeContext(event.data.sessionId, resumeUrl)
+        }
+
         on<Close.SessionReset> {
-            session.update { null }
+            resumeContext.value = null
         }
```

Comment:
I think it might be necessary to reset the context on some other `Close` events, e.g. `UserClose` (resuming after the user closed the gateway seems wrong). I'm new to gateway code, so I might be wrong here, what do you think?

Reply:
I do assume that this is for the resumeGatewayUrl new property but why not just store that new URL in place of the old gateway url

Reply:
we need to keep the old one because we have to connect to it when we don't resume a session but instead start a new one

Reply:
Oh, this looks fine then